File: src/sdk/stablecoin.ts
// Solana Stablecoin Standard - TypeScript SDK
// Layer 1: Base SDK + Layer 2: Compliance Module + Layer 3: SSS-2 Preset

import {
  Connection,
  PublicKey,
  Keypair,
  Transaction,
  TransactionInstruction,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
} from '@solana/web3.js';
import { Program, AnchorProvider, web3 } from '@project-serum/anchor';

// ID do programa (gerado automaticamente pelo Anchor)
const STABLECOIN_PROGRAM_ID = new PublicKey('STABLE111111111111111111111111111111111');

// ============================================================================
// Interfaces
// ============================================================================

export interface StablecoinConfig {
  name: string;
  symbol: string;
  uri: string;
  decimals: number;
  enable_permanent_delegate: boolean;
  enable_transfer_hook: boolean;
  default_account_frozen: boolean;
  mint_authority: PublicKey;
  freeze_authority: PublicKey;
}

export interface BlacklistEntry {
  address: PublicKey;
  is_blacklisted: boolean;
  timestamp: number;
}

export interface MintToParams {
  to: PublicKey;
  amount: number;
}

export interface FreezeAccountParams {
  account: PublicKey;
}

export interface ModifyBlacklistParams {
  address: PublicKey;
}

export interface SeizeTokensParams {
  from: PublicKey;
  treasury: PublicKey;
  amount: number;
}

export interface UpdateAuthorityParams {
  newAuthority: PublicKey;
}

export interface InitializeWithTransferHookParams {
  name: string;
  symbol: string;
  uri: string;
  decimals: number;
  transferHookProgram?: PublicKey; // NOVO: Transfer hook program (opcional)
}

// ============================================================================
// StablecoinSDK Class
// ============================================================================

export class StablecoinSDK {
  public program: Program;
  public programId: PublicKey;

  constructor(
    provider: AnchorProvider,
    idl: any,
    programId: PublicKey = STABLECOIN_PROGRAM_ID
  ) {
    this.programId = programId;
    this.program = new Program(idl, programId, provider);
  }

  /**
   * Initialize SSS-2 Compliant Stablecoin
   * Creates mint, sets up authorities, and configures compliance features
   */
  async initialize(config: StablecoinConfig): Promise<string> {
    const mint = Keypair.generate();
    const [stablecoinConfig] = await PublicKey.findProgramAddress(
      [Buffer.from('stablecoin'), mint.publicKey.toBuffer()],
      this.programId
    );

    const [mintAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('mint'), mint.publicKey.toBuffer()],
      this.programId
    );

    const [freezeAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('freeze'), mint.publicKey.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .initialize(
        config.name,
        config.symbol,
        config.uri,
        config.decimals,
        config.enable_permanent_delegate,
        config.default_account_frozen
      )
      .accounts({
        stablecoinConfig,
        mint: mint.publicKey,
        mintAuthority: config.mint_authority,
        freezeAuthority: config.freeze_authority,
        authority: this.program.provider.wallet.publicKey,
        payer: this.program.provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .signers([mint])
      .rpc();

    console.log('Stablecoin initialized:', tx);
    return tx;
  }

  /**
   * Initialize stablecoin WITH transfer hook (NOVO)
   * Allows specifying a separate transfer hook program for blacklist enforcement
   */
  async initializeWithTransferHook(
    config: StablecoinConfig & {
      transferHookProgram?: PublicKey
    }
  ): Promise<string> {
    const mint = Keypair.generate();
    const [stablecoinConfig] = await PublicKey.findProgramAddress(
      [Buffer.from('stablecoin'), mint.publicKey.toBuffer()],
      this.programId
    );

    const [mintAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('mint'), mint.publicKey.toBuffer()],
      this.programId
    );

    const [freezeAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('freeze'), mint.publicKey.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .initialize(
        config.name,
        config.symbol,
        config.uri,
        config.decimals,
        config.enable_permanent_delegate,
        config.default_account_frozen,
        config.transferHookProgram || null // NOVO: Hook program opcional
      )
      .accounts({
        stablecoinConfig,
        mint: mint.publicKey,
        mintAuthority: config.mint_authority,
        freezeAuthority: config.freeze_authority,
        authority: this.program.provider.wallet.publicKey,
        payer: this.program.provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .signers([mint])
      .rpc();

    console.log('Stablecoin initialized with hook:', config.transferHookProgram);
    return tx;
  }

  /**
   * Mint new tokens to a recipient account
   * Only callable by mint authority
   */
  async mintTo(params: MintToParams): Promise<string> {
    const [mintAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('mint'), params.to.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .mintTo(new web3.BN(params.amount))
      .accounts({
        mint: params.to,
        mintAuthority,
        to: params.to,
        payer: this.program.provider.wallet.publicKey,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .rpc();

    return tx;
  }

  /**
   * Freeze a token account (compliance)
   * Uses Token-2022 Freeze extension
   */
  async freezeAccount(params: FreezeAccountParams): Promise<string> {
    const [freezeAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('freeze'), params.account.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .freezeAccount()
      .accounts({
        mint: params.account,
        tokenAccount: params.account,
        freezeAuthority,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .rpc();

    return tx;
  }

  /**
   * Thaw a frozen token account
   * Uses Token-2022 Thaw extension
   */
  async thawAccount(params: FreezeAccountParams): Promise<string> {
    const [freezeAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('freeze'), params.account.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .thawAccount()
      .accounts({
        mint: params.account,
        tokenAccount: params.account,
        freezeAuthority,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .rpc();

    return tx;
  }

  /**
   * Add an address to blacklist (compliance)
   * Stores blacklist status in PDA for Transfer Hook
   */
  async addToBlacklist(params: ModifyBlacklistParams): Promise<string> {
    const [blacklistEntry] = await PublicKey.findProgramAddress(
      [Buffer.from('blacklist'), params.address.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .addToBlacklist(params.address)
      .accounts({
        blacklistEntry,
        authority: this.program.provider.wallet.publicKey,
        payer: this.program.provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .rpc();

    return tx;
  }

  /**
   * Remove an address from blacklist
   */
  async removeFromBlacklist(params: ModifyBlacklistParams): Promise<string> {
    const [blacklistEntry] = await PublicKey.findProgramAddress(
      [Buffer.from('blacklist'), params.address.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .removeFromBlacklist(params.address)
      .accounts({
        blacklistEntry,
        authority: this.program.provider.wallet.publicKey,
        payer: this.program.provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .rpc();

    return tx;
  }

  /**
   * Seize tokens from a blacklisted account (compliance)
   * Freeze + Transfer to treasury
   */
  async seizeTokens(params: SeizeTokensParams): Promise<string> {
    const [freezeAuthority] = await PublicKey.findProgramAddress(
      [Buffer.from('authority'), Buffer.from('freeze'), params.from.toBuffer()],
      this.programId
    );

    const tx = await this.program.methods
      .seizeTokens(new web3.BN(params.amount))
      .accounts({
        mint: params.from,
        from: params.from,
        fromAuthority: this.program.provider.wallet.publicKey,
        treasury: params.treasury,
        freezeAuthority,
        tokenProgram: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
      })
      .rpc();

    return tx;
  }

  /**
   * Update mint authority (multi-sig compatible)
   */
  async updateMinter(params: UpdateAuthorityParams): Promise<string> {
    const tx = await this.program.methods
      .updateMinter(params.newAuthority)
      .accounts({
        authority: this.program.provider.wallet.publicKey,
      })
      .rpc();

    return tx;
  }

  /**
   * Update freeze authority (multi-sig compatible)
   */
  async updateFreezer(params: UpdateAuthorityParams): Promise<string> {
    const tx = await this.program.methods
      .updateFreezer(params.newAuthority)
      .accounts({
        authority: this.program.provider.wallet.publicKey,
      })
      .rpc();

    return tx;
  }

  /**
   * Check if an address is blacklisted
   * Returns true if blacklisted, false otherwise
   */
  async isBlacklisted(address: PublicKey): Promise<boolean> {
    const [blacklistEntry] = await PublicKey.findProgramAddress(
      [Buffer.from('blacklist'), address.toBuffer()],
      this.programId
    );

    try {
      const account = await this.program.account.blacklistEntry.fetch(blacklistEntry);
      return account.isBlacklisted;
    } catch (error) {
      // Account doesn't exist = not blacklisted
      return false;
    }
  }

  /**
   * Get stablecoin configuration
   */
  async getConfig(mint: PublicKey): Promise<StablecoinConfig> {
    const [stablecoinConfig] = await PublicKey.findProgramAddress(
      [Buffer.from('stablecoin'), mint.toBuffer()],
      this.programId
    );

    const account = await this.program.account.stablecoinConfig.fetch(stablecoinConfig);

    return {
      name: account.name,
      symbol: account.symbol,
      uri: account.uri,
      decimals: account.decimals,
      enable_permanent_delegate: account.enablePermanentDelegate,
      enable_transfer_hook: account.enableTransferHook,
      default_account_frozen: account.defaultAccountFrozen,
      mint_authority: new PublicKey(account.mintAuthority),
      freeze_authority: new PublicKey(account.freezeAuthority),
    };
  }
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Create a new StablecoinSDK instance
 */
export function createStablecoinSDK(
  connection: Connection,
  wallet: any,
  programId?: PublicKey
): StablecoinSDK {
  const provider = new AnchorProvider(connection, wallet, {
    commitment: 'confirmed',
  });
  AnchorProvider.setProvider(provider);

  // Load IDL (this would be generated by Anchor build)
  const idl = {
    // IDL would be loaded here
    version: '0.1.0',
    name: 'stablecoin',
    instructions: [],
    accounts: [],
    types: [],
  };

  return new StablecoinSDK(provider, idl, programId);
}

/**
 * Example: Initialize a new stablecoin
 */
export async function exampleInitialize(): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana; // Phantom wallet

  const sdk = createStablecoinSDK(connection, wallet);

  const config: StablecoinConfig = {
    name: 'My Stablecoin',
    symbol: 'MSTBL',
    uri: 'https://example.com/metadata.json',
    decimals: 9,
    enable_permanent_delegate: true,
    enable_transfer_hook: true,
    default_account_frozen: false,
    mint_authority: wallet.publicKey,
    freeze_authority: wallet.publicKey,
  };

  const tx = await sdk.initialize(config);
  console.log('Stablecoin initialized:', tx);
}

/**
 * Example: Initialize stablecoin WITH transfer hook (NOVO)
 */
export async function exampleInitializeWithTransferHook(): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana; // Phantom wallet
  const hookProgramId = new PublicKey('hookProg11111111111111111111'); // Simulado para teste

  const sdk = createStablecoinSDK(connection, wallet);

  const config: StablecoinConfig = {
    name: 'My Stablecoin',
    symbol: 'MSTBL',
    uri: 'https://example.com/metadata.json',
    decimals: 9,
    enable_permanent_delegate: true,
    enable_transfer_hook: true,
    default_account_frozen: false,
    mint_authority: wallet.publicKey,
    freeze_authority: wallet.publicKey,
    transferHookProgram: hookProgramId, // NOVO: Especificar hook program
  };

  const tx = await sdk.initializeWithTransferHook(config);
  console.log('Stablecoin initialized with hook:', tx);
}

/**
 * Example: Mint tokens to a recipient
 */
export async function exampleMintTo(to: PublicKey, amount: number): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana;

  const sdk = createStablecoinSDK(connection, wallet);

  const tx = await sdk.mintTo({ to, amount });
  console.log('Tokens minted:', tx);
}

/**
 * Example: Freeze an account
 */
export async function exampleFreezeAccount(account: PublicKey): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana;

  const sdk = createStablecoinSDK(connection, wallet);

  const tx = await sdk.freezeAccount({ account });
  console.log('Account frozen:', tx);
}

/**
 * Example: Add address to blacklist
 */
export async function exampleAddToBlacklist(address: PublicKey): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana;

  const sdk = createStablecoinSDK(connection, wallet);

  const tx = await sdk.addToBlacklist({ address });
  console.log('Address blacklisted:', tx);
}

/**
 * Example: Seize tokens from a blacklisted account
 */
export async function exampleSeizeTokens(
  from: PublicKey,
  treasury: PublicKey,
  amount: number
): Promise<void> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana;

  const sdk = createStablecoinSDK(connection, wallet);

  const tx = await sdk.seizeTokens({ from, treasury, amount });
  console.log('Tokens seized:', tx);
}

/**
 * Example: Check if address is blacklisted
 */
export async function exampleIsBlacklisted(address: PublicKey): Promise<boolean> {
  const connection = new Connection('https://api.devnet.solana.com');
  const wallet = window.solana;

  const sdk = createStablecoinSDK(connection, wallet);

  const isBlacklisted = await sdk.isBlacklisted(address);
  console.log('Is blacklisted:', isBlacklisted);

  return isBlacklisted;
}

// ============================================================================
// Exports
// ============================================================================

export { STABLECOIN_PROGRAM_ID };

File: programs/solana-stablecoin-standard/src/stablecoin.rs
// Solana Stablecoin Standard - SSS-2 Compliant Implementation
// Migrating to Token-2022 with Transfer Hook

use anchor_lang::prelude::*;
use anchor_spl::token_2022::{
    self,
    extensions::{
        transfer_hook::TransferHook,
        mint::Mint,
    },
    state::Mint,
    instruction::TransferChecked,
};

// Re-export error codes from lib.rs
use super::StablecoinError;

#[program]
pub mod stablecoin {
    use super::*;

    /// Initialize SSS-2 Compliant Stablecoin with Token-2022
    /// Creates mint with transfer hook for blacklist enforcement
    pub fn initialize(
        ctx: Context<Initialize>,
        name: String,
        symbol: String,
        uri: String,
        decimals: u8,
    ) -> Result<()> {
        let stablecoin_config = &mut ctx.accounts.stablecoin_config;
        
        // Validate string lengths
        require!(name.len() <= 32, StablecoinError::NameTooLong);
        require!(symbol.len() <= 16, StablecoinError::SymbolTooLong);
        require!(uri.len() <= 256, StablecoinError::UriTooLong);
        
        stablecoin_config.name = name;
        stablecoin_config.symbol = symbol;
        stablecoin_config.uri = uri;
        stablecoin_config.decimals = decimals;
        stablecoin_config.mint_authority = ctx.accounts.authority.key();
        stablecoin_config.freeze_authority = ctx.accounts.authority.key();
        stablecoin_config.bump = ctx.bumps.stablecoin_config;

        msg!("Stablecoin initialized with Token-2022: {}", stablecoin_config.symbol);
        Ok(())
    }
        stablecoin_config.uri = uri;
        stablecoin_config.decimals = decimals;
        stablecoin_config.mint_authority = ctx.accounts.authority.key();
        stablecoin_config.freeze_authority = ctx.accounts.authority.key();
        stablecoin_config.bump = ctx.bumps.stablecoin_config;

        msg!("Stablecoin initialized with Token-2022: {}", stablecoin_config.symbol);
        Ok(())
    }

    /// Mint new tokens to a recipient account
    /// Uses Token-2022 Mint extension
    pub fn mint_to(
        ctx: Context<MintTo>,
        amount: u64,
    ) -> Result<()> {
        let cpi_accounts = token_2022::instruction::MintTo {
            token_program: ctx.accounts.token_program.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            to: ctx.accounts.to.to_account_info(),
            authority: ctx.accounts.authority.to_account_info(),
        };
        
        token_2022::mint_to(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                cpi_accounts,
            ),
            amount,
        )?;
        
        msg!("Minted {} tokens to {}", amount, ctx.accounts.to.key());
        Ok(())
    }

    /// Freeze a token account (compliance)
    /// Uses Token-2022 Freeze extension
    pub fn freeze_account(ctx: Context<FreezeAccount>) -> Result<()> {
        let cpi_accounts = token_2022::instruction::FreezeAccount {
            token_program: ctx.accounts.token_program.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            account: ctx.accounts.token_account.to_account_info(),
            authority: ctx.accounts.authority.to_account_info(),
        };
        
        token_2022::freeze_account(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                cpi_accounts,
            ),
        )?;
        
        msg!("Froze account {}", ctx.accounts.token_account.key());
        Ok(())
    }

    /// Thaw a frozen token account
    /// Uses Token-2022 Thaw extension
    pub fn thaw_account(ctx: Context<ThawAccount>) -> Result<()> {
        let cpi_accounts = token_2022::instruction::ThawAccount {
            token_program: ctx.accounts.token_program.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            account: ctx.accounts.token_account.to_account_info(),
            authority: ctx.accounts.authority.to_account_info(),
        };
        
        token_2022::thaw_account(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                cpi_accounts,
            ),
        )?;
        
        msg!("Thawed account {}", ctx.accounts.token_account.key());
        Ok(())
    }

    /// Add an address to blacklist (compliance)
    /// Stores blacklist status in PDA for Transfer Hook
    pub fn add_to_blacklist(
        ctx: Context<ModifyBlacklist>, 
        address: Pubkey
    ) -> Result<()> {
        let blacklist_entry = &mut ctx.accounts.blacklist_entry;
        
        // Verify authority
        require!(
            ctx.accounts.authority.key() == ctx.accounts.stablecoin_config.mint_authority,
            StablecoinError::UnauthorizedAccess
        );

        blacklist_entry.address = address;
        blacklist_entry.is_blacklisted = true;
        blacklist_entry.timestamp = Clock::get()?.unix_timestamp;

        msg!("Added address {} to blacklist (for Transfer Hook)", address);
        Ok(())
    }

    /// Remove an address from blacklist
    pub fn remove_from_blacklist(
        ctx: Context<ModifyBlacklist>, 
        address: Pubkey
    ) -> Result<()> {
        let blacklist_entry = &mut ctx.accounts.blacklist_entry;
        
        // Verify authority
        require!(
            ctx.accounts.authority.key() == ctx.accounts.stablecoin_config.mint_authority,
            StablecoinError::UnauthorizedAccess
        );

        blacklist_entry.address = address;
        blacklist_entry.is_blacklisted = false;
        blacklist_entry.timestamp = Clock::get()?.unix_timestamp;

        msg!("Removed address {} from blacklist", address);
        Ok(())
    }

    /// Seize tokens from a blacklisted account (compliance)
    /// Freeze + Transfer to treasury
    pub fn seize_tokens(
        ctx: Context<SeizeTokens>,
        amount: u64,
    ) -> Result<()> {
        // First freeze the account
        let freeze_cpi = token_2022::instruction::FreezeAccount {
            token_program: ctx.accounts.token_program.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            account: ctx.accounts.from.to_account_info(),
            authority: ctx.accounts.authority.to_account_info(),
        };
        
        token_2022::freeze_account(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                freeze_cpi,
            ),
        )?;

        // Then transfer to treasury
        let transfer_cpi = token_2022::instruction::TransferChecked {
            token_program: ctx.accounts.token_program.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            from: ctx.accounts.from.to_account_info(),
            to: ctx.accounts.treasury.to_account_info(),
            authority: ctx.accounts.from_authority.to_account_info(),
            amount: ctx.accounts.amount,
            decimals: ctx.accounts.mint.decimals,
        };
        
        token_2022::transfer_checked(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                transfer_cpi,
            ),
            amount,
        )?;

        msg!("Seized {} tokens from {} to treasury", amount, ctx.accounts.from.key());
        Ok(())
    }

    /// Update mint authority (multi-sig compatible)
    pub fn update_minter(
        ctx: Context<UpdateAuthority>,
        new_mint_authority: Pubkey,
    ) -> Result<()> {
        let stablecoin_config = &mut ctx.accounts.stablecoin_config;
        
        require!(
            ctx.accounts.authority.key() == stablecoin_config.mint_authority,
            StablecoinError::UnauthorizedAccess
        );

        stablecoin_config.mint_authority = new_mint_authority;
        msg!("Updated mint authority to {}", new_mint_authority);
        Ok(())
    }

    /// Update freeze authority (multi-sig compatible)
    pub fn update_freezer(
        ctx: Context<UpdateAuthority>,
        new_freeze_authority: Pubkey,
    ) -> Result<()> {
        let stablecoin_config = &mut ctx.accounts.stablecoin_config;
        
        require!(
            ctx.accounts.authority.key() == stablecoin_config.freeze_authority,
            StablecoinError::UnauthorizedAccess
        );

        stablecoin_config.freeze_authority = new_freeze_authority;
        msg!("Updated freeze authority to {}", new_freeze_authority);
        Ok(())
    }
}

// ============================================================================
// Contexts (Account Validation)
// ============================================================================

#[derive(Accounts)]
#[instruction(name: String, symbol: String, uri: String, decimals: u8, transfer_hook_program: Option<Pubkey>)]
pub struct Initialize<'info> {
    #[account(
        init,
        payer = payer,
        space = 8 + 32 + 16 + 256 + 1 + 32 + 32 + 1, // discriminator + strings + decimals + authorities + bump
        seeds = [b"stablecoin", mint.key().as_ref()],
        bump
    )]
    pub stablecoin_config: Account<'info, StablecoinConfig>,

    #[account(
        init,
        payer = payer,
        space = 32 + 1, // program key + discriminator
        seeds = [b"transfer_hook", transfer_hook_program.as_ref()],
        bump
    )]
    pub transfer_hook_program: UncheckedAccount<'info>,

    #[account(
        init,
        payer = payer,
        mint::decimals = 9,
        mint::authority = authority,
        // NOVO: Configurar mint com hook program se fornecido
        extension::transfer_hook::TransferHook {
            program: transfer_hook_program,
            authority: mint_authority,
        },
    )]
    pub mint: Account<'info, Mint>,

    #[account(mut)]
    pub transfer_hook_program: UncheckedAccount<'info>,

    #[account(mut)]
    pub authority: Signer<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, token_2022::ID>,
}

#[derive(Accounts)]
pub struct MintTo<'info> {
    #[account(mut)]
    pub mint: Account<'info, Mint>,

    #[account(
        init,
        payer = payer,
        token::mint = mint,
        token::authority = authority,
    )]
    pub to: Account<'info, TokenAccount>,

    #[account(mut)]
    pub authority: Signer<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, token_2022::ID>,
}

#[derive(Accounts)]
pub struct FreezeAccount<'info> {
    #[account(mut)]
    pub mint: Account<'info, Mint>,

    #[account(mut)]
    pub token_account: Account<'info, TokenAccount>,

    #[account(mut)]
    pub authority: Signer<'info>,

    pub token_program: Program<'info, token_2022::ID>,
}

#[derive(Accounts)]
pub struct ThawAccount<'info> {
    #[account(mut)]
    pub mint: Account<'info, Mint>,

    #[account(mut)]
    pub token_account: Account<'info, TokenAccount>,

    #[account(mut)]
    pub authority: Signer<'info>,

    pub token_program: Program<'info, token_2022::ID>,
}

#[derive(Accounts)]
pub struct ModifyBlacklist<'info> {
    #[account(
        init_if_needed,
        payer = payer,
        space = 8 + 32 + 1 + 8, // discriminator + address + bool + timestamp
        seeds = [b"blacklist", address.as_ref()],
        bump
    )]
    pub blacklist_entry: Account<'info, BlacklistEntry>,

    pub stablecoin_config: Account<'info, StablecoinConfig>,

    #[account(mut)]
    pub authority: Signer<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct SeizeTokens<'info> {
    #[account(mut)]
    pub mint: Account<'info, Mint>,

    #[account(mut)]
    pub from: Account<'info, TokenAccount>,

    #[account(mut)]
    pub from_authority: Signer<'info>,

    #[account(mut)]
    pub treasury: Account<'info, TokenAccount>,

    #[account(mut)]
    pub authority: Signer<'info>,

    pub token_program: Program<'info, token_2022::ID>,
}

#[derive(Accounts)]
pub struct UpdateAuthority<'info> {
    #[account(mut)]
    pub stablecoin_config: Account<'info, StablecoinConfig>,

    #[account(mut)]
    pub authority: Signer<'info>,
}

// ============================================================================
// Data Structures
// ============================================================================

#[account]
pub struct StablecoinConfig {
    pub name: String,
    pub symbol: String,
    pub uri: String,
    pub decimals: u8,
    pub mint_authority: Pubkey,
    pub freeze_authority: Pubkey,
    pub bump: u8,
}

#[account]
pub struct BlacklistEntry {
    pub address: Pubkey,
    pub is_blacklisted: bool,
    pub timestamp: i64,
}
